sample-schemas/demo_schema.erl

```diff
@@ -0,0 +1,57 @@
+-module(demo_schema).
+
+-include_lib("typerefl/include/types.hrl").
+
+-type birthdays() :: [#{m := 1..12, d := 1..31}].
+
+-reflect_type([birthdays/0]).
+
+-export([namespaces/0, fields/1]).
+
+namespaces() -> ["foo", "a.b", "x.y"].
+
+fields("foo") ->
+    [ {"setting", fun setting/1}
+    , {"endpoint", fun endpoint/1}
+    , {"greet", fun greet/1}
+    , {"numbers", fun numbers/1}
+    , {"ref_x_y", fun ref_x_y/1}
+    ];
+
+fields("a.b") ->
+    [ {"birthdays", fun birthdays/1}
+    , {"some_int", fun int/1}
+    ];
+
+fields("x.y") ->
+    [ {"some_int", fun int/1}
+    ].
+
+setting(mapping) -> "app_foo.setting";
+setting(type) -> string();
+setting(override_env) -> "MY_OVERRIDE";
+setting(_) -> undefined.
+
+endpoint(mapping) -> "app_foo.endpoint";
+endpoint(type) -> typerefl:ip4_address();
+endpoint(_) -> undefined.
+
+greet(mapping) -> "app_foo.greet";
+greet(type) -> typerefl:regexp_string("^hello$");
+greet(_) -> undefined.
+
+numbers(mapping) -> "app_foo.numbers";
+numbers(type) -> list(integer());
+numbers(_) -> undefined.
+
+birthdays(mapping) -> "a.b.birthdays";
+birthdays(type) -> birthdays();
+birthdays(_) -> undefined.
+
+int(mapping) -> "a.b.some_int";
+int(type) -> integer();
+int(_) -> undefined.
+
+ref_x_y(mapping) -> undefined;
```

Comment:
I might have gave a bad example above. sorry.
we actually need a demo which shows how it works to map an object to a target,
so, please change it to some mapping target.
the `{"some_int", fun int/1}` field has no need for a mapping target, that's fine.

then in the test case, we'd expect something like:
```
% ?_assertEqual([], F("foo.ref_x_y={some_int = 1}")) % not this, but below:
?_assertEqual([{["foo", "ref_x_y"], #{"some_int" => 1}}], F("foo.ref_x_y={some_int = 1}"))
% or maybe 
?_assertEqual([{["foo", "ref_x_y"], [{"some_int", 1}]}], F("foo.ref_x_y={some_int = 1}"))
```

Reply:
the output of `map` API should be the list of mapped values. Since `foo.ref_x_y` is a raw key, `translate` API should be responsible for it when we want to get `"foo.ref_x_y={some_int = 1}"` ?

Reply:
yes, if `mapping` clause returns `undefined` then it should not be mapped.

what I am after is (if not to change `ref_x_y`, then maybe add another one):
`ref_j_k(mapping) -> "app_foo.jkopts";`
so the raw values parsed from hocon is mapped to `jkopts`.
i.e. a one to one mapping from `ref_j_k.*` to `app_foo.jkopts.*`.
with this, we do not have to write mapping for each field,
hence we can use built in types like `fun int/1` for the fields.

Reply:
I see. demo added